posthog/tasks/exporter.py

```diff
@@ -120,6 +128,61 @@ def _export_to_png(exported_asset: ExportedAsset) -> None:
             driver.close()
 
 
+def encode(obj: object) -> str:
+    if isinstance(obj, uuid.UUID):
+        return str(obj)
+
+    if isinstance(obj, datetime.datetime):
+        return obj.isoformat()
+
+    return str(obj)
+
+
+def stage_results_to_object_storage(
+    day_filter: Filter, exported_asset: ExportedAsset, temporary_file: IO, write_headers: bool
+) -> None:
+    # load results
+    result = query_events_list(
+        filter=day_filter,
+        team=exported_asset.team,
+        request_get_query_dict=exported_asset.export_context.get("request_get_query_dict"),
+        order_by=exported_asset.export_context.get("order_by"),
+        action_id=exported_asset.export_context.get("action_id"),
+        limit=100_000_000,  # what limit do we want?! None ¯\_(ツ)_/¯
+    )
+    if write_headers:
+        temporary_file.write(gzip.compress(f"{','.join(result[0].keys())}\n".encode("utf-8")))
+
+    contents = []
+    for values in [row.values() for row in result]:
+        line = [encode(v) for v in values]
+        contents.append(",".join(line))
+
+    temporary_file.write(gzip.compress("\n".join(contents).encode("utf-8")))
+
+
+def concat_results_in_object_storage(temporary_file: IO, exported_asset: ExportedAsset) -> str:
+    object_path = f"/exports/csvs/team-{exported_asset.team.id}/task-{exported_asset.id}/{UUIDT()}"
+    temporary_file.seek(0)
+    object_storage.write(object_path, temporary_file.read())
```

Comment:
It's a step in that direction (and I improved it by thinking about it more carefully so I could respond 🚀 )

We load the entire dataset from ClickHouse so have to have enough memory to hold that and then we make a new representation of that (the CSV rows)

Rather than hold that new representation in memory we write each line as we come to it to the temporary file.

Then we pass the read method of the file to gzip to compress it hoping that it does something clever :)

And then pass that compressed result to the S3 write. Boto3 then decides whether to upload in one go or to do multipart uploads - which is brill cos I was worried we'd have to write that

So, it could probably be cleverer but since export is pretty much entirely broken now I thought this would be a good step along